From 5c4987f8b15cbf809fa80990872340466eacd835 Mon Sep 17 00:00:00 2001
From: ejgoldik <148076610+ejgoldik@users.noreply.github.com>
Date: Wed, 16 Jul 2025 13:53:59 +0300
Subject: [PATCH 52/80] d4xx-compat: Add error for hwmc response len 0 +
 replace hard-coded addresses (#285)

* Added err and dmesg in ds5_get_hwmc if response length is 0
* Added comments and replaced hard coded HWMC addresses
---
 drivers/media/i2c/d4xx-compat/d4xx.c | 58 +++++++++++++++-------------
 1 file changed, 32 insertions(+), 26 deletions(-)

diff --git a/drivers/media/i2c/d4xx-compat/d4xx.c b/drivers/media/i2c/d4xx-compat/d4xx.c
index d292916c..6b474aa5 100644
--- a/drivers/media/i2c/d4xx-compat/d4xx.c
+++ b/drivers/media/i2c/d4xx-compat/d4xx.c
@@ -2245,18 +2245,24 @@ static int ds5_get_hwmc(struct ds5 *state, unsigned char *data,
 	}
 
 	ret = regmap_raw_read(state->regmap, DS5_HWMC_RESP_LEN,
-			&tmp_len, sizeof(tmp_len));
+			&tmp_len, sizeof(tmp_len)); /* Read response length */
 	if (ret)
 		return -EBADMSG;
 
 	if (tmp_len > cmdDataLen)
 		return -ENOBUFS;
 
+	if (tmp_len == 0) {
+		dev_err(&state->client->dev,
+			"%s(): HWMC response length is 0\n", __func__);
+		return -ENODATA;
+	}
+
 	dev_dbg(&state->client->dev,
 			"%s(): HWMC read len: %d, lrs_len: %d\n",
 			__func__, tmp_len, tmp_len - 4);
 
-	ds5_raw_read_with_check(state, DS5_HWMC_DATA, data, tmp_len);
+	ds5_raw_read_with_check(state, DS5_HWMC_DATA, data, tmp_len); /* Read response data */
 	if (dataLen)
 		*dataLen = tmp_len;
 	return ret;
@@ -2272,8 +2278,8 @@ static int ds5_send_hwmc(struct ds5 *state,
 			__func__, cmd->header, cmd->magic_word, cmd->opcode,
 			cmdLen,	cmd->param1, cmd->param2, cmd->param3, cmd->param4);
 
-	ds5_raw_write_with_check(state, DS5_HWMC_DATA, cmd, cmdLen);
-	
+	ds5_raw_write_with_check(state, DS5_HWMC_DATA, cmd, cmdLen); /* Write command data */
+
 	ds5_write_with_check(state, DS5_HWMC_EXEC, 0x01); /* execute cmd */
 
 	return 0;
@@ -2286,7 +2292,7 @@ static int ds5_set_calibration_data(struct ds5 *state,
 	int retries = 10;
 	u16 status = 2;
 
-	ds5_raw_write_with_check(state, DS5_HWMC_DATA, cmd, length);
+	ds5_raw_write_with_check(state, DS5_HWMC_DATA, cmd, length); /* Write command data */
 
 	ds5_write_with_check(state, DS5_HWMC_EXEC, 0x01); /* execute cmd */
 	do {
@@ -2691,12 +2697,12 @@ static int ds5_get_calibration_data(struct ds5 *state, enum table_id id,
 
 	memcpy(cmd, &get_calib_data, sizeof(get_calib_data));
 	cmd->param1 = id;
-	ds5_raw_write_with_check(state, 0x4900, cmd, sizeof(struct hwm_cmd));
-	ds5_write_with_check(state, 0x490c, 0x01); /* execute cmd */
+	ds5_raw_write_with_check(state, DS5_HWMC_DATA, cmd, sizeof(struct hwm_cmd)); /* Write command data */
+	ds5_write_with_check(state, DS5_HWMC_EXEC, 0x01); /* execute cmd */
 	do {
 		if (retries != 3)
 			msleep_range(10);
-		ret = ds5_read(state, 0x4904, &status);
+		ret = ds5_read(state, DS5_HWMC_STATUS, &status);
 	} while (ret && retries-- && status != 0);
 
 	if (ret || status != 0) {
@@ -2708,11 +2714,11 @@ static int ds5_get_calibration_data(struct ds5 *state, enum table_id id,
 	}
 
 	// get table length from fw
-	ret = regmap_raw_read(state->regmap, 0x4908,
-			&table_length, sizeof(table_length));
+	ret = regmap_raw_read(state->regmap, DS5_HWMC_RESP_LEN,
+			&table_length, sizeof(table_length)); /* Read response length */
 
 	// read table
-	ds5_raw_read_with_check(state, 0x4900, cmd->Data, table_length);
+	ds5_raw_read_with_check(state, DS5_HWMC_DATA, cmd->Data, table_length); /* Read table data */
 
 	// first 4 bytes are opcode HWM, not part of calibration table
 	memcpy(table, cmd->Data + 4, length);
@@ -2729,13 +2735,13 @@ static int ds5_gvd(struct ds5 *state, unsigned char *data)
 	u8 retries = 3;
 
 	memcpy(&cmd, &gvd, sizeof(gvd));
-	ds5_raw_write_with_check(state, 0x4900, &cmd, sizeof(cmd));
-	ds5_write_with_check(state, 0x490c, 0x01); /* execute cmd */
+	ds5_raw_write_with_check(state, DS5_HWMC_DATA, &cmd, sizeof(cmd)); /* Write command data */
+	ds5_write_with_check(state, DS5_HWMC_EXEC, 0x01); /* execute cmd */
 	do {
 		if (retries != 3)
 			msleep_range(10);
 
-		ret = ds5_read(state, 0x4904, &status);
+		ret = ds5_read(state, DS5_HWMC_STATUS, &status);
 	} while (ret && retries-- && status != 0);
 
 	if (ret || status != 0) {
@@ -2745,8 +2751,8 @@ static int ds5_gvd(struct ds5 *state, unsigned char *data)
 		return status;
 	}
 
-	ret = regmap_raw_read(state->regmap, 0x4908, &length, sizeof(length));
-	ds5_raw_read_with_check(state, 0x4900, data, length);
+	ret = regmap_raw_read(state->regmap, DS5_HWMC_RESP_LEN, &length, sizeof(length)); /* Read response length */
+	ds5_raw_read_with_check(state, DS5_HWMC_DATA, data, length); /* Read response data */
 
 	return ret;
 }
@@ -2865,19 +2871,19 @@ static int ds5_g_volatile_ctrl(struct v4l2_ctrl *ctrl)
 		//       1. prepare and send command
 		//       2. send command
 		//       3. execute command
-		//       4. wait for ccompletion
-		ret = regmap_raw_write(state->regmap, 0x4900,
+		//       4. wait for completion
+		ret = regmap_raw_write(state->regmap, DS5_HWMC_DATA, /* Write command data */
 				log_prepare, sizeof(log_prepare));
 		if (ret < 0)
 			return ret;
 
-		ret = regmap_raw_write(state->regmap, 0x490C,
-				&execute_cmd, sizeof(execute_cmd));
+		ret = regmap_raw_write(state->regmap, DS5_HWMC_EXEC,
+				&execute_cmd, sizeof(execute_cmd)); /* execute cmd */
 		if (ret < 0)
 			return ret;
 
 		for (i = 0; i < DS5_MAX_LOG_POLL; i++) {
-			ret = regmap_raw_read(state->regmap, 0x4904,
+			ret = regmap_raw_read(state->regmap, DS5_HWMC_STATUS,
 					&data, sizeof(data));
 			dev_dbg(&state->client->dev, "%s(): log ready 0x%x\n",
 				 __func__, data);
@@ -2891,7 +2897,7 @@ static int ds5_g_volatile_ctrl(struct v4l2_ctrl *ctrl)
 //		if (i == DS5_MAX_LOG_POLL)
 //			return -ETIMEDOUT;
 
-		ret = regmap_raw_read(state->regmap, 0x4908, &data, sizeof(data));
+		ret = regmap_raw_read(state->regmap, DS5_HWMC_RESP_LEN, &data, sizeof(data)); /* Read response length */
 		dev_dbg(&state->client->dev, "%s(): log size 0x%x\n", __func__, data);
 		if (ret < 0)
 			return ret;
@@ -2899,7 +2905,7 @@ static int ds5_g_volatile_ctrl(struct v4l2_ctrl *ctrl)
 			return 0;
 		if (data > 1024)
 			return -ENOBUFS;
-		ret = regmap_raw_read(state->regmap, 0x4900,
+		ret = regmap_raw_read(state->regmap, DS5_HWMC_DATA,
 				ctrl->p_new.p_u8, data);
 		break;
 	case DS5_CAMERA_DEPTH_CALIBRATION_TABLE_GET:
@@ -6227,9 +6233,9 @@ static int ds5_dfu_switch_to_dfu(struct ds5 *state)
 	int i = DS5_START_MAX_COUNT;
 	u16 status;
 
-	ds5_raw_write_with_check(state, 0x4900,
-			&cmd_switch_to_dfu, sizeof(cmd_switch_to_dfu));
-	ds5_write_with_check(state, 0x490c, 0x01); /* execute cmd */
+	ds5_raw_write_with_check(state, DS5_HWMC_DATA,
+			&cmd_switch_to_dfu, sizeof(cmd_switch_to_dfu)); /* Write command data */
+	ds5_write_with_check(state, DS5_HWMC_EXEC, 0x01); /* execute cmd */
 	/*Wait for DFU fw to boot*/
 	do {
 		msleep_range(DS5_START_POLL_TIME*10);
-- 
2.25.1

